### PR TITLE
Architecture: Static Typing for Native Invocations

### PR DESCRIPTION
**Pull request by talekhinezh**
* Support for static typing for native function and method invocations, skipping serialization/deserialization when not required
* Further separated concerns between interpreters and kernel, the kernel is much less dependent on `ScryptoValue` now for example

### CODE REVIEW

File: radix-engine/src/engine/kernel.rs [hunk 1]
```diff
@@ -1615,3 +1375,261 @@ where
...
+        }
+
+        node_refs_to_copy.insert(RENodeId::Global(GlobalAddress::Resource(RADIX_TOKEN)));
+        node_refs_to_copy.insert(RENodeId::Global(GlobalAddress::System(EPOCH_MANAGER)));
+        node_refs_to_copy.insert(RENodeId::Global(GlobalAddress::Resource(
+            ECDSA_SECP256K1_TOKEN,
+        )));
+        node_refs_to_copy.insert(RENodeId::Global(GlobalAddress::Resource(
+            EDDSA_ED25519_TOKEN,
+        )));
```
Comment (iamyulong) [inline radix-engine/src/engine/kernel.rs:1623]: Why do we still need this? To allow scrypto direct access to these entities without ref passing?

File: radix-engine/src/engine/interpreters/native_interpreter.rs [hunk 2]
```diff
@@ -74,74 +75,145 @@ impl Into<ApplicationError> for EpochManagerError {
...
+    + Invokable<ResourceManagerGetMetadataInvocation>
+    + Invokable<ResourceManagerGetResourceTypeInvocation>
+    + Invokable<ResourceManagerGetTotalSupplyInvocation>
+    + Invokable<ResourceManagerUpdateMetadataInvocation>
+    + Invokable<ResourceManagerUpdateNonFungibleDataInvocation>
+    + Invokable<ResourceManagerNonFungibleExistsInvocation>
+    + Invokable<ResourceManagerGetNonFungibleInvocation>
+    + Invokable<ResourceManagerSetResourceAddressInvocation>
+    + Invokable<EpochManagerGetCurrentEpochInvocation>
+    + Invokable<EpochManagerSetEpochInvocation>
```
Comment (iamyulong) [inline radix-engine/src/engine/interpreters/native_interpreter.rs:136]: I think we will need more thoughts on forks very soon. IE if we add more invokable in the future, we will need combination of invokables.

File: radix-engine/src/model/nodes/transaction_processor.rs [hunk 3]
```diff
@@ -228,596 +220,473 @@ impl TransactionProcessor {
         }
     }
 
-    pub fn static_main<'s, Y, R>(
-        func: TransactionProcessorFunction,
-        args: ScryptoValue,
+    pub fn static_main<'a, Y>(
```
Comment (iamyulong) [inline radix-engine/src/model/nodes/transaction_processor.rs:223]: MInor:  Can we rename `static_main` to `run`, so to help me forget main dispatching?

File: scrypto/src/component/component.rs [hunk 4]
```diff
@@ -138,17 +141,6 @@ impl BorrowedGlobalComponent {
         let state: DataRef<ComponentInfoSubstate> = pointer.get();
         state.blueprint_name.clone()
     }
-
-    pub fn add_access_check(&mut self, access_rules: AccessRules) -> &mut Self {
```
Comment (iamyulong) [inline scrypto/src/component/component.rs:142]: No more `add_access_check` for globalized components. I like it!

File: radix-engine/src/engine/native_wrapper.rs [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+            let invocation: PackagePublishInvocation = scrypto_decode(&args)
+                .map_err(|e| RuntimeError::KernelError(KernelError::DecodeError(e)))?;
+            system_api
+                .invoke(invocation)
+                .map(|a| ScryptoValue::from_typed(&a))
+        }
+    }
+}
+
+// TODO: Cleanup
```
Comment (iamyulong) [inline radix-engine/src/engine/native_wrapper.rs:60]: Just double check that the clean-up is to decouple `Wasm::NativeInvocation` and `Transaction::NativeInvocation` right? :)

File: radix-engine/src/engine/interpreters/native_interpreter.rs [hunk 6]
```diff
@@ -74,74 +75,145 @@ impl Into<ApplicationError> for EpochManagerError {
...
-        match fn_identifier {
-            NativeFunction::TransactionProcessor(func) => {
-                TransactionProcessor::static_main(func, input, system_api).map_err(|e| e.into())
-            }
-            NativeFunction::Package(func) => {
-                Package::static_main(func, input, system_api).map_err(|e| e.into())
-            }
-            NativeFunction::ResourceManager(func) => {
-                ResourceManager::static_main(func, input, system_api).map_err(|e| e.into())
+pub trait InvokableNative<'a>:
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/engine/interpreters/native_interpreter.rs:78]: why not? `TransactionProcessorRunInvocation<'a>` requires it?

File: radix-engine/src/engine/kernel.rs [hunk 7]
```diff
@@ -644,453 +638,233 @@ where
...
+    N: NativeInvocation,
+{
+    fn invoke(&mut self, invocation: N) -> Result<<N as Invocation>::Output, RuntimeError> {
+        for m in &mut self.modules {
+            m.pre_sys_call(
+                &self.current_frame,
+                &mut self.heap,
+                &mut self.track,
+                SysCallInput::Invoke {
+                    name: format!("{:?}", invocation),
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/engine/kernel.rs:782]: Yup agreed, will leave this to a future PR though.

File: radix-engine/src/engine/kernel.rs (hunk 1 above)
Comment (talekhinezh) [reply radix-engine/src/engine/kernel.rs:1623]: Yup we have a couple of examples which still do this.

File: radix-engine/src/model/nodes/transaction_processor.rs (hunk 3 above)
Comment (talekhinezh) [reply radix-engine/src/model/nodes/transaction_processor.rs:223]: yup!

File: scrypto/src/component/component.rs (hunk 4 above)
Comment (talekhinezh) [reply scrypto/src/component/component.rs:142]: I think this is temporary and will need to add back in the future :)

File: radix-engine/src/engine/native_wrapper.rs (hunk 5 above)
Comment (talekhinezh) [reply radix-engine/src/engine/native_wrapper.rs:60]: yup exactly

File: radix-engine/src/engine/interpreters/native_interpreter.rs (hunk 2 above)
Comment (iamyulong) [reply radix-engine/src/engine/interpreters/native_interpreter.rs:136]: I was thinking what would happen if we've added extra invocation and we need to run past transaction without the added invocations.

Comment (talekhinezh) [reply radix-engine/src/engine/interpreters/native_interpreter.rs:136]: ah, gotcha. Good point! I think one easy approach may be separate traits for each fork version?

**Pull request merged**